File: cpp/lib/tools/func.cpp
/* =====================================================================================
                     The UQ Toolkit (UQTk) version 3.0
                     Copyright (2016) Sandia Corporation
                     http://www.sandia.gov/UQToolkit/

     Copyright (2016) Sandia Corporation. Under the terms of Contract DE-AC04-94AL85000
     with Sandia Corporation, the U.S. Government retains certain rights in this software.

     This file is part of The UQ Toolkit (UQTk)

     UQTk is free software: you can redistribute it and/or modify
     it under the terms of the GNU Lesser General Public License as published by
     the Free Software Foundation, either version 3 of the License, or
     (at your option) any later version.

     UQTk is distributed in the hope that it will be useful,
     but WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
     GNU Lesser General Public License for more details.

     You should have received a copy of the GNU Lesser General Public License
     along with UQTk.  If not, see <http://www.gnu.org/licenses/>.

     Questions? Contact Bert Debusschere <bjdebus@sandia.gov>
     Sandia National Laboratories, Livermore, CA, USA
===================================================================================== */
/// \file func.cpp
/// \brief Implements several functions of form \f$y=f(\lambda;x)\f$

#include <math.h>
#include <cfloat>
#include <assert.h>

#include "func.h"

#include "gen_defs.h"
#include "PCSet.h"
#include "error_handlers.h"
#include "arrayio.h"
#include "arraytools.h"


// Proportionality function f(p,x) = p * x
Array2D<double> Func_Prop(Array2D<double>& p, Array2D<double>& x,void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int pdim=p.YSize();
	int xdim=x.YSize();

	assert(pdim==1);
	assert(xdim==1);

	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=p(ip,0)*x(ix,0);

	return y;
}

// Quadratic proportionality f(p,x) = p1 * x + p2 * x^2
Array2D<double> Func_PropQuad(Array2D<double>& p, Array2D<double>& x,void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int pdim=p.YSize();
	int xdim=x.YSize();

	assert(pdim==2);
	assert(xdim==1);
	
	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=p(ip,0)*x(ix,0)+p(ip,1)*x(ix,0)*x(ix,0);

	return y;
}

// Exponential function f(p,x) = exp ( p1 + p2 * x )
Array2D<double> Func_Exp(Array2D<double>& p, Array2D<double>& x,void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int pdim=p.YSize();
	int xdim=x.YSize();

	assert(pdim==2);
	assert(xdim==1);
	
	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=exp(p(ip,0)+p(ip,1)*x(ix,0));

	return y;
}

// Quadratic exponential function f(p,x) = exp ( p1 + p2 * x + p3 * x^2 )
Array2D<double> Func_ExpQuad(Array2D<double>& p, Array2D<double>& x,void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int pdim=p.YSize();
	int xdim=x.YSize();

	assert(pdim==3);
	assert(xdim==1);
	
	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=exp(p(ip,0)+p(ip,1)*x(ix,0)+p(ip,2)*x(ix,0)*x(ix,0));

	return y;
}

// Constant function f(p,x) = p
Array2D<double> Func_Const(Array2D<double>& p, Array2D<double>& x,void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int pdim=p.YSize();
	int xdim=x.YSize();

	assert(pdim==1);
	
	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=p(ip,0);

	return y;
}

// Linear function f(p,x) = p1 + p2 * x
Array2D<double> Func_Linear(Array2D<double>& p, Array2D<double>& x,void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int pdim=p.YSize();
	int xdim=x.YSize();

	assert(pdim==2);
	assert(xdim==1);
	
	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=p(ip,0)+p(ip,1)*x(ix,0);

	return y;
}

// Black-box function f(p,x), expects a script bb.x which reads p.dat and x.dat and writs y.dat
Array2D<double> Func_BB(Array2D<double>& p, Array2D<double>& x,void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int pdim=p.YSize();
	int xdim=x.YSize();

	
	Array2D<double> y;

	write_datafile(p,"p.dat"); //np,pdim
    write_datafile(x,"x.dat"); //nx,xdim
    system("./bb.x");
    read_datafileVS(y,"y.dat"); //np,nx

    CHECKEQ(y.XSize(),np);
    CHECKEQ(y.YSize(),nx);

	return y;
}

// Heat transfer example function f(p,x) = T0 + (x * dw) / (Aw * p)
Array2D<double> Func_HT1(Array2D<double>& p, Array2D<double>& x,void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int pdim=p.YSize();
	int xdim=x.YSize();

	assert(pdim==1);
	assert(xdim==1);
	
	Array2D<double> y(np,nx);

	double To=273.;
   	double Aw=0.04;
    double dw=0.1;	
	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=x(ix,0)*dw/(Aw*p(ip,0))+To;

	return y;
}

// Heat transfer example function f(p,x) = p2 + (x * Q) / (Aw * p1)
Array2D<double> Func_HT2(Array2D<double>& p, Array2D<double>& x,void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int pdim=p.YSize();
	int xdim=x.YSize();

	assert(pdim==2);
	assert(xdim==1);
	
	Array2D<double> y(np,nx);

	double Aw=0.04;
    //double kw=1.0;
    double Q=20.0;
	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
		    y(ip,ix)=x(ix,0)*Q/(Aw*p(ip,0))+p(ip,1);

	return y;
}

// Fractional power example function f(p,x) = p1 + p2 * x + p3 * x^2 + p4 * (x + 1)^3.5 
Array2D<double> Func_FracPower(Array2D<double>& p, Array2D<double>& x,void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int pdim=p.YSize();
	int xdim=x.YSize();

	assert(pdim==4);
	assert(xdim==1);
	

	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=p(ip,0)+p(ip,1)*x(ix,0)+p(ip,2)*pow(x(ix,0),2.)+p(ip,3)*pow(x(ix,0)+1.,3.5);

	return y;
}

// Exponential example function for quick sketch f(p,x) = p2 * exp( x * p1 ) - 2
Array2D<double> Func_ExpSketch(Array2D<double>& p, Array2D<double>& x,void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int pdim=p.YSize();
	int xdim=x.YSize();

	assert(pdim==2);
	assert(xdim==1);
	
	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=p(ip,1)*exp(x(ix,0)*p(ip,0))-2.0;

	return y;
}

// Function that returns all inputs as a vector f(p,x_i) = p_i for i=1,2,..,nx
Array2D<double> Func_Inputs(Array2D<double>& p, Array2D<double>& x,void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int pdim=p.YSize();
	int xdim=x.YSize();
	
	assert(pdim==nx);
	assert(xdim==1);
	
	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=p(ip,ix);

	return y;
}

// LU expansion f(p,x) = \sum_k p_k \Psi_k(x)
Array2D<double> Func_PCl(Array2D<double>& p, Array2D<double>& x,void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int pdim=p.YSize();
	int xdim=x.YSize();

	Array2D<int> mindexx;

	string pctype="LU";
    read_datafileVS(mindexx,"mindexx.dat");

    // Size checks
    CHECKEQ(mindexx.XSize(),pdim);
    CHECKEQ(mindexx.YSize(),xdim);

        
	PCSet surrmodel("NISPnoq",mindexx,pctype);
        
    Array2D<double> pcinput=x;
    
    Array2D<double> y(np,nx);

    for(int ip=0;ip<np;ip++){
    	Array1D<double> cf_this;
    	getRow(p,ip,cf_this);
        Array1D<double> moutput;
        surrmodel.EvalPCAtCustPoints(moutput,pcinput,cf_this);
        
        for(int ix=0;ix<nx;ix++)
            y(ip,ix)=moutput(ix);
    }

	return y;
}

// LU expansion f(p,x) = \sum_k c_k \Psi(p,x)
Array2D<double> Func_PCx(Array2D<double>& p, Array2D<double>& x,void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int pdim=p.YSize();
	int xdim=x.YSize();

	Array2D<int> mindexpx;
	string pctype="LU";
    read_datafileVS(mindexpx,"mindexpx.dat");

    CHECKEQ(mindexpx.YSize(),pdim+xdim);

    Array1D<double> pccfpx(mindexpx.XSize());
    read_datafile_1d(pccfpx,"pccfpx.dat");
        
	PCSet surrmodel("NISPnoq",mindexpx,pctype);

    Array2D<double> pcinput(nx*np,pdim+xdim);
    for(int ip=0;ip<np;ip++){
	    for(int ix=0;ix<nx;ix++){
            for(int j=0;j<pdim;j++)
                pcinput(ip+ix*np,j)=p(ip,j);
            for(int j=0;j<xdim;j++)
                pcinput(ip+ix*np,j+pdim)=x(ix,j);
        }
    }
    Array1D<double> model_samples_planar;
    surrmodel.EvalPCAtCustPoints(model_samples_planar,pcinput,pccfpx);

	Array2D<double> y(np,nx);

    for (int ip=0;ip<np;ip++)
        for (int ix=0;ix<nx;ix++)
            y(ip,ix)=model_samples_planar(ip+ix*np);

	return y;
}

// LU expansion f(p,x_i) = \sum_k c_k \Psi_k(p_i) when the multiindex is shared for all i
Array2D<double> Func_PC(Array2D<double>& p, Array2D<double>& x,void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int pdim=p.YSize();
	int xdim=x.YSize();

	Array2D<int> mindexp;

	string pctype="LU";
    read_datafileVS(mindexp,"mindexp.dat");
    CHECKEQ(mindexp.YSize(),pdim);

    Array2D<double> pccf_all(mindexp.XSize(),nx);
    read_datafile(pccf_all,"pccf_all.dat");
        
	PCSet surrmodel("NISPnoq",mindexp,pctype);
        
    Array2D<double> pcinput=p;
    
    Array2D<double> y(np,nx);

    for(int ix=0;ix<nx;ix++){
        Array1D<double> pccf_this;
        getCol(pccf_all,ix,pccf_this);
        Array1D<double> moutput;
        surrmodel.EvalPCAtCustPoints(moutput,pcinput,pccf_this);
        
        for(int ip=0;ip<np;ip++)
            y(ip,ix)=moutput(ip);
    }


	return y;
}

// LU expansion f(p,x_i) = \sum_k c_k \Psi_k(p_i) when multiindex can be different for each i
Array2D<double> Func_PCs(Array2D<double>& p, Array2D<double>& x,void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int pdim=p.YSize();
	int xdim=x.YSize();

	string pctype="LU";

	Array1D<Array2D<int> > mindices(nx);
    Array1D<Array1D<double> > pccfs(nx);

   	for (int ix=0;ix<nx;ix++){
        char filename[25];
        int nn;
        
        nn=sprintf(filename,"mindexp.%d.dat",ix);
        read_datafileVS(mindices(ix), filename);
        assert(mindices(ix).YSize()==pdim);

        nn=sprintf(filename,"pccfp.%d.dat",ix);
        pccfs(ix).Resize(mindices(ix).XSize());
        
        read_datafile_1d(pccfs(ix), filename);

	}
   
        
        
    Array2D<double> pcinput=p;
    
    Array2D<double> y(np,nx);

    for(int ix=0;ix<nx;ix++){
		PCSet surrmodel("NISPnoq",mindices(ix),pctype);

        Array1D<double> moutput;
        surrmodel.EvalPCAtCustPoints(moutput,pcinput,pccfs(ix));
        
        for(int ip=0;ip<np;ip++)
            y(ip,ix)=moutput(ip);
    }


	return y;
}

File: PyUQTk/utils/func.py
#!/usr/bin/env python
#=====================================================================================
#                     The UQ Toolkit (UQTk) version 3.0
#                     Copyright (2016) Sandia Corporation
#                     http://www.sandia.gov/UQToolkit/
#
#     Copyright (2016) Sandia Corporation. Under the terms of Contract DE-AC04-94AL85000
#     with Sandia Corporation, the U.S. Government retains certain rights in this software.
#
#     This file is part of The UQ Toolkit (UQTk)
#
#     UQTk is free software: you can redistribute it and/or modify
#     it under the terms of the GNU Lesser General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     UQTk is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU Lesser General Public License for more details.
#
#     You should have received a copy of the GNU Lesser General Public License
#     along with UQTk.  If not, see <http://www.gnu.org/licenses/>.
#
#     Questions? Contact Bert Debusschere <bjdebus@sandia.gov>
#     Sandia National Laboratories, Livermore, CA, USA
#===================================================================================== 

"""
Generic tools for evaluation of standard functions 
and their integrals
"""

try:
    import numpy as np
except ImportError:
    "Need numpy"


from math import *
import random as rnd
import itertools

###################################################################################################


def func(xdata,model,func_params):
    """Generic function evaluator.
    Note:
        * Currently only genz functions are implemented.
        * Note that conventional Genz arguments are in [0,1], here the expected input is on [-1,1]
    Arguments:
        * xdata       : Nxd numpy array of input, should be in [-1,1]^d
        * model       : Model name, options are 'genz_osc', 'genz_exp', 'genz_cont', 'genz_gaus', 'genz_cpeak', 'genz_ppeak'
        * func_params : Auxiliary parameters
                      : For genz functions, an array of size d+1, the first entry being the shift, and the rest of the entries are the weights.
                      : See UQTk Manual for Genz formulae.
    Returns:
        * ydata       : An array of outputs of size N.
    """

    # Get the input size
    sam=xdata.shape[0]
    dim=xdata.shape[1]
    
    # Check the function types and evaluate
    if model == 'genz_osc':
        xdata=0.5*(xdata+1.)
        ydata=empty((sam,))
        gcf=func_params[1:]
        xtmp=dot(xdata,gcf)
        for j in range(sam):
            ydata[j]=cos(2.*pi*func_params[0]+xtmp[j])
    
    elif model == 'genz_exp':
        xdata=0.5*(xdata+1.)
        ydata=empty((sam,))
        ww=func_params[0]
        gcf=func_params[1:]
        
        xtmp=dot(xdata-ww,gcf)
        for j in range(sam):
            ydata[j]=exp(xtmp[j])
    
    elif model == 'genz_cont':
        xdata=0.5*(xdata+1.)
        ydata=empty((sam,))
        ww=func_params[0]
        gcf=func_params[1:]
        
        xtmp=dot(abs(xdata-ww),gcf)
        for j in range(sam):
            ydata[j]=exp(-xtmp[j])
    
    elif model == 'genz_gaus':
        xdata=0.5*(xdata+1.)
        ydata=empty((sam,))
        ww=func_params[0]
        gcf=func_params[1:]
        
        xtmp=dot((xdata-ww)*(xdata-ww),gcf*gcf)
        for j in range(sam):
            ydata[j]=exp(-xtmp[j])
    
    elif model == 'genz_cpeak':
        xdata=0.5*(xdata+1.)
        ydata=empty((sam,))
        #ww=param[0]
        gcf=func_params[1:]
        
        xtmp=1.+(dot(xdata,gcf)) #use abs if defined on [-1,1]
        for j in range(sam):
            ydata[j]=exp(-(dim+1.)*log(xtmp[j]))
    
    elif model == 'genz_ppeak':
        xdata=0.5*(xdata+1.)
        ydata=empty((sam,))
        ww=func_params[0]
        gcf=func_params[1:]
        
        for j in range(sam):
            prod=1.
            for i in range(dim):
                prod = prod / (1./(gcf[i]**2.)+(xdata[j,i]-ww)**2.)
            ydata[j]=prod

    return ydata

def integ_exact(model,func_params):
    """Generic function integral evaluator.
    Note:
        * Currently only genz functions are implemented.
        * Note that conventional Genz arguments are in [0,1], here the expected input is on [-1,1]
    Arguments:
        * model       : Model name, options are 'genz_osc', 'genz_exp', 'genz_cont', 'genz_gaus', 'genz_cpeak', 'genz_ppeak'
        * func_params : Auxiliary parameters
                      : For genz functions, an array of size d+1, the first entry being the shift, and the rest of the entries are the weights.
                      : See UQTk Manual for Genz integral formulae.
    Returns:
        * integ_ex    : A real number that is the integral over [-1,1]^d
    """


    if (model=='genz_osc'):
        gcf=func_params
        dim=gcf.shape[0]-1
        integ_ex=cos(2.*pi*gcf[0]+0.5*sum(gcf[1:]))
        for i in range(1,dim+1):
            integ_ex*=(2.*sin(gcf[i]/2.)/gcf[i])
    elif (model=='genz_exp'):
        gcf=func_params
        dim=gcf.shape[0]-1
        integ_ex=1.
        for i in range(1,dim+1):
            at1=exp(-gcf[i]*gcf[0])
            at2=exp(gcf[i]*(1.-gcf[0]))
            integ_ex*=((at2-at1)/(gcf[i]))
    elif (model=='genz_cont'):
        gcf=func_params
        dim=gcf.shape[0]-1
        integ_ex=1.
        for i in range(1,dim+1):
            integ_ex*= ((2.-exp(gcf[i]*(-gcf[0]))-exp(gcf[i]*(gcf[0]-1.)))/gcf[i])
    elif (model=='genz_gaus'):
        gcf=func_params
        dim=gcf.shape[0]-1
        integ_ex=1.
        for i in range(1,dim+1):
            at1=math.erf(-gcf[i]*gcf[0])
            at2=math.erf(gcf[i]*(1.-gcf[0]))
            integ_ex*=((at2-at1)*sqrt(pi)/(2.*gcf[i]))
    elif (model=='genz_cpeak'):
        gcf=func_params
        dim=gcf.shape[0]-1
        numer=0.0
        count=1
        denom=1.
        for i in range(1,dim+1):
            comb=list(itertools.combinations(range(1,dim+1),i))
            for j in range(len(comb)):
                assert(i==len(comb[j]))
                #print i,j,pow(-1,i)
                numer+=(pow(-1,i)/(1.+sum(gcf[list(comb[j])])))
                count+=1
            denom*=(i*gcf[i])
        #print count, numer
        integ_ex=(1.+numer)/denom
    elif (model=='genz_ppeak'):
        gcf=func_params
        dim=gcf.shape[0]-1
        integ_ex=1.
        for i in range(1,dim+1):
            at1=arctan(-gcf[i]*gcf[0])
            at2=arctan(gcf[i]*(1.-gcf[0]))
            integ_ex*=(gcf[i]*(at2-at1))

    return integ_ex

